Decentralized project integration

rkzel opened the pull request:
NewProject panel has been connected to the contract via api.addProject function. New field, 'decoupled', is used to discern between GH and native projects.

DecoratedRepos context is where most interesting changes happened: new function has been added to pull metadata from IPFS - previously there was only one function there collecting information about GH projects directly from GH via Apollo client. Instead of adding IPFS to the same function, new function seemed to make sense, because they are only superficially similar - there were subtle differences, enough to warrant keeping the functionality separate. Also, in case there are more sources of repos/projects/issues added, it shall be done in the same manner.

What is working now:
- creating and displaying native and GH projects.
- removing both types.

What is not working:
- editing Projects,
- ProjectsDetail tab (that's waiting for: PR #1923  Projects: improve querying behavior).

topocount committed:
feat(projects): add decoupled project functionality

- add decentralized project entries
- add and update issues directly without bounties
- enable multi-fulfillment bounties

BREAKING CHANGE: Projects contract updates

topocount committed:
feat(projects): add decoupled project functionality

- add decentralized project entries
- add and update issues directly without bounties
- enable multi-fulfillment bounties

BREAKING CHANGE: Projects contract updates

topocount committed:
fix(projects): enable radspec for external function

- `fulfillBounty` radspec message now shows up in external tx panel
- remove extraneous logging from deploy script
- remove unneeded functions from Bounties interface

topocount committed:
fix(projects): enable radspec for external function

- `fulfillBounty` radspec message now shows up in external tx panel
- remove extraneous logging from deploy script
- remove unneeded functions from Bounties interface

Commit:
refactor(projects): use standard apollo gql syntax

Commit:
refactor(projects): lean on Apollo for pagination

* import `ApolloProvider` from `@apollo/react-hooks` so we don't need to
  re-define `client` in ProjectDetail
* use `fetchMore` returned from `useQuery` call instead of manually
  fetching more
* show "loading" animation below current results – much smoother loading
  of more items!
* fetch 25 issues at a time, instead of 100 – makes whole experience
  much faster

This last one currently makes filtering even more dorky, until we
start modifying the query instead of doing filtering locally

Commit:
fix(projects): sorting issues re-queries GitHub

Note that GitHub does not allow sorting by "Name". GitHub's sort
options:

* CREATED_AT
* UPDATED_AT
* COMMENTS (sorts by comment count)

I changed the drop-down to only show the first two

Commit:
refactor(projects): use GitHub `search` query

This will allow us to support free-form queries when the user types in
the search field. Paging issues within a `repository` lookup does not.

Note that this method of searching does not allow sorting by creation
date: https://help.github.com/en/github/searching-for-information-on-github/sorting-search-results

After playing around with the available sorting options, I thought
"updated" seemed the only useful one, in the context of the Projects
app. I couldn't figure out what "author-date" was actually sorting on,
but it doesn't seem to be the creation date of the issue.

Commit:
refactor(projects): rm unused filter keys

Commit:
refactor(projects): fix Query & Mutation components

Switch these to use hooks so that they work with the ApolloProvider from
'@apollo/react-hooks'

Commit:
refactor(projects): useBountyIssues in ProjectDetail

This also fixes an issue where these bountyIssues were not restricted to
only those for the selected project

Commit:
refactor(projects): filter&sort bountyIssues locally

bountyIssues only live locally. As more logic moves to the GitHub query,
we want to make sure we keep around logic that allows us to provide a
good experience when filtering for issues by bounty status.

This adds back local sorting, but it only gets used when filtering for
bounty status. It also cleans up filtering logic to make it work for the
new desired paradigm.

rkzel committed:
feat: new decentralized projects can now be created

NewProject panel has been connected to the contract via api.addProject
function. New field, 'decoupled', is used to discern between GH
and native projects.

DecoratedRepos context is where most interesting changes happened:
new function has been added to pull metadata from IPFS - previously
there was only one function there collecting information about GH
projects directly from GH via Apollo client. Instead of adding IPFS
to the same function, new function seemed to make sense, because
they are only superficially similar - there were subtle differences,
enough to warrant keeping the functionality separate. Also, in case
there are more sources of repos/projects/issues added, it shall
be done in the same manner.

What is working now
- creating and displaying native and GH projects.
- removing both types.

What is not working
- editing Projects,
- ProjectsDetail tab (that's waiting for PR 1923, improve querying behavior).

Quazia commented on apps/projects/app/components/Panel/NewProject/NewProject.js:
```diff
@@ -113,110 +121,187 @@ const NewProject = () => {
+        if (data && data.viewer) {
+          const reposDownloaded = filterAlreadyAdded(data.viewer.repositories.edges)
+          const visibleRepos = filter ? filterByName(reposDownloaded) : reposDownloaded
+          // eslint-disable-next-line react/prop-types
```
Why are we disabling the prop-types here?

rkzel committed:
refactor(filterbar): removed counters from filters dropdowns

Quazia commented on apps/projects/app/components/Panel/NewProject/NewProject.js:
```diff
@@ -113,110 +121,187 @@ const NewProject = () => {
-          </Query>
+  return (
+    <>
+      {github.status !== 'authenticated' ? (
```
As it is now, this isn't accessible since the _initial_ splash screen requires auth. We need to go straight into project view.

topocount committed:
feat(projects): add projects updateability

Quazia committed:
feat: integrate with add/update for decoupled repo

rkzel committed:
refactor: moved labels and milestones retrieval to DecoratedRepo

rkzel committed:
refactor: filtering moved to GH server side

rkzel committed:
fix: useMemo was unreachable for IssueDetail route

rkzel committed:
fix: show more repos button

PeterMPhillips committed:
fix(dot-voting): dark mode support for voting bars

Quazia committed:
fix: decoupled bool for setRepo in createProject

topocount committed:
fix(vote slider): add theme for slider background

rkzel committed:
feat: text filter uses GH server-side filtering

Quazia committed:
fix: resole issue with key vs repoId in General.js

Quazia committed:
Merge branch 'dev' of https://github.com/AutarkLabs/planning-suite into issue-querying

Quazia committed:
fix: toggle off decoupled projects

topocount committed:
chore(projects): integrate `setIssue` panel call

rkzel committed:
feat: removed disfunctional milestones filtering

rkzel committed:
Merge branch 'issue-querying' of github.com:AutarkLabs/open-enterprise into issue-querying

* 'issue-querying' of github.com:AutarkLabs/open-enterprise:
  fix: toggle off decoupled projects
  fix(vote slider): add theme for slider background
  fix(dot-voting): dark mode support for voting bars
  fix: change radspec for newVote
  fix:  Very small dot-voting participation rounded
  fix(recovery script): error on invalid allocations period length
  refactor: feature-flip, ThematicBreak
  fix(projects): use main component with emptywrapper
  fix(projects): ignore bounties made before Nov 26
  fix: hidden InfoBox, adjusted ScrollList height, fixed intendation
  refactor: github query component and post-review fixes
  feat: added UI for native projects to NewProject panel
  fix(tests): fix tests failing because running in parallel

topocount committed:
chore(projects): add contract state handling for decoupled issues

Quazia committed:
Merge pull request #1923 from AutarkLabs/issue-querying

Projects: improve querying behavior

Quazia committed:
Merge branch 'decoupled-projects' of https://github.com/AutarkLabs/planning-suite into decentralized-project-integration

topocount committed:
chore(projects): integrate Edit Project Panel

topocount committed:
chore(projects): add closepanel in decoupled issue api call

Quazia committed:
chore(projects): decentralized project detail view

Quazia committed:
chore(projects):  decoupled issues in detail view

Quazia committed:
Merge branch 'decentralized-project-integration' of https://github.com/AutarkLabs/planning-suite into decentralized-project-integration

topocount committed:
chore(projects): decoupled bounties now showing up in the bounties tab

decoupled Bounties are now registering in both the bounties tab and project details
Remaining kinks for me to work out:
- get bounty workflow completing for decoupled bounties. There are a few
small hangups related to how repoId's get transalated between ascii and
hex
- get issue details work with decoupled bounties. Graphql queries are
still attemping in issue details and they must be disabled when the repo
is decoupled
- ensure github bounty flows are unchanged. I think I've been steering
clear of altering any github bounty logic pathways but this obviously
needs to be tests before this goes into review

topocount committed:
chore(projects): integrate decoupled issues with bounty kaban

- bounty flow works for decoupled issues
- created fallback for generating repo hex ids since it's now stored but
was not in the past
Other Issues to address
- back button for decoupled issues
- small formatting things for decoupled issue detail
- possibly adding in a markdown editor for issue detail creation
- empty card for project detail, which has a new issue button

topocount committed:
chore(projects): completed integrating decoupled projects

topocount committed:
chore(projects): address @ottodevs contract review

- remove magic numbers
- add in safemath where necessary
- clean up contract documentation

topocount committed:
Merge branch 'dev' into decentralized-project-integration

ottodevs commented on apps/projects/app/components/Panel/NewProject/NewProject.js:
```diff
@@ -3,16 +3,19 @@ import PropTypes from 'prop-types'
 
-const disableDecoupledProjects = true
+
+const disableDecoupledProjects = false
```
This seems not needed anymore, as it cannot mutate in runtime, we can get rid of it unless we plan to do something with it in the future.
(Also remove condition in line 297)

ottodevs commented on apps/projects/app/components/Content/IssueDetail/index.js:
```diff
@@ -46,37 +55,70 @@ Wrap.propTypes = {
 
+  const storedIssue = useMemo(() => {
+    return issues.find(i => issueId === i.data.issueId )
+  }, [issues])
```
We may need `issueId` also here, I suggest adding the react-hooks linting rules to better detect these cases, as we have in about app and in the client, so maybe not needed to change here, (as it does not seem to explicitly break the `storedIssue` value), but better when we add the rule in a separate commit.

ottodevs commented on apps/projects/app/components/Content/ProjectDetail.js:
```diff
@@ -227,43 +225,86 @@ class ProjectDetail extends React.PureComponent {
-    const allIssues = [ ...cachedIssues, ...data.repository.issues.nodes ]
+    let dataSource
+    let pageInfo
+    if (data && !this.props.repo.decoupled) {
```
This condition does not seem necessary as line 228 already limits the `data` value to non-decoupled repos

ottodevs commented on apps/projects/app/components/Content/ProjectDetail.js:
```diff
@@ -277,54 +318,102 @@ class ProjectDetail extends React.PureComponent {
+    repo: repo.decoupled ? repo.id : `${repo.metadata.owner}/${repo.metadata.name}`,
+    search: '',
+    sort: 'updated-desc',
+    owner: repo.decoupled ? 'AutarkLabs' : repo.metadata.owner,
```
Are we sure we need to hardcode `AutarkLabs` here?

ottodevs commented on apps/projects/app/components/Content/ProjectDetail.js:
```diff
@@ -277,54 +318,102 @@ class ProjectDetail extends React.PureComponent {
+    search: '',
+    sort: 'updated-desc',
+    owner: repo.decoupled ? 'AutarkLabs' : repo.metadata.owner,
+    name: repo.decoupled ? 'open-enterprise' : repo.metadata.name,
```
Are we sure we want to hardcode the `open-enteprise` value?

ottodevs commented on apps/projects/app/components/Panel/FundIssues/FundIssues.js:
```diff
@@ -48,7 +48,10 @@ const errorMessages = {
 
 const bountiesFor = ({ bountySettings, issues, tokens }) => issues.reduce(
   (bounties, issue) => {
+    const data = issue.repository.decoupled ? issue : undefined
```
call here just to ensure that `undefined` is the appropriate type for non-decoupled repos data field

ottodevs commented on apps/projects/app/components/Panel/NewIssue/NewIssue.js:
```diff
@@ -103,66 +112,125 @@ class NewIssue extends React.PureComponent {
+        },
+        repository: reposManaged[selectedProject - 1],
+        labels: { totalCount: 0, edges: Array(0) },
+        milestone: null,
```
It could make sense to get rid of the `milestone` field if we don't want to use it anymore, it was removed from the filterbar but still present across the code, and apparently always being assigned a "null" value, what is the purpose of having a value that will always be null? This may happen for other fields here.

ottodevs commented on apps/projects/app/components/Panel/NewIssue/NewIssue.js:
```diff
@@ -103,66 +112,125 @@ class NewIssue extends React.PureComponent {
+            required
+          />
+        </Field>
+        {selectedProject > 0 && (reposManaged[selectedProject - 1].decoupled ? (
```
No need to change right now, but just FYI we can use the dropdown placeholder to avoid these index modification hacks that are error-prone:
https://github.com/aragon/aragon-ui/tree/master/src/components/DropDown#placeholder

ottodevs commented on apps/projects/app/context/DecoratedRepos.js:
```diff
@@ -1,13 +1,25 @@
       url
       description
+      owner { login }
+      labels(first: 100) {
```
We should add pagination as we have in the issues query, as this is omitting lot of repositories.

topocount replied:
I'm curious what this might look like, since the `useDecoratedRepos` hook that implements this query is used only to populate repos that are already added. We;re just not paginating the labels on these repos we know we already want to query. I've created [AUT-92](https://linear.app/issue/AUT-92/add-pagination-to-repo-query) to discuss this further

ottodevs commented on apps/projects/app/utils/shapes.js:
```diff
@@ -30,9 +30,30 @@ const issueShape = PropTypes.shape({
 
 const userGitHubShape = PropTypes.shape({
-  avatarUrl: PropTypes.string.isRequired,
+  avatarUrl: PropTypes.string,
```
It seems removing `isRequired` will hide potential warning messages, because at least `PanelComponents.js`->`Avatar` component is using this `prop` without any explicit check for its value, so it is actually a required prop unless we add any kind of check to that component (maybe there are other components involved as well)

ottodevs reviewed (dismissed):
Overall I didn't find code that is clearly wrong or causing malfunction of the app, just asking some minor questions, that are not necessarily involving further changes to this PR, but would be good to clarify in case they could hide some potential caveats.

Besides that, I detected some issues while testing the app locally,  the biggest problem I experienced was not being able to create decoupled issues nor create permissions to do so

I could not assign the needed permission to create decoupled issues, as it is not figuring in the permissions panel:

![image](https://user-images.githubusercontent.com/5030059/75475617-d94e7780-5977-11ea-8f1a-1adae0a68c34.png)

But still showing up as unassigned in the general permissions list:
![image](https://user-images.githubusercontent.com/5030059/75475720-0733bc00-5978-11ea-8e02-80611f648916.png)

ottodevs commented:
Fixes AUT-19

topocount committed:
chore(projects): address @ottodevs review

- fix user proptypes and incorporate decoupled user
- remove milestone property
- change issue data fall back from `undefined` to an empty object
- remove default repo name and owner properties set for gql query
- short-circuit repo gql query if the repo is decoupled
- remove unnecessary decoupled check
- add issueId to the update array for `useMemo`

topocount committed:
Merge remote-tracking branch 'origin/dev' into decentralized-project-integration

ottodevs reviewed (approved):
Just an overall recommendation, PLEASE AVOID FORCE-PUSHING PR changes once there is an active review, because it becomes impossible to track the changes this way...  Specially with such huge PR. 

Alternatively, there is a squash option when merging where we can squash every commit, so Force-pushing does not seem to offer any kind of advantage over that option.

Besides that personal advice/request, this seems ok now, great changes!

- Decoupled projects and their issues worked fine in my environment 
- Permissions look correct now, the "Create and update projects issues" one is finally showing up

As long as we consider the coverage decrease as acceptable for now, this looks ready to merge for me.
